keras_nlp/layers/preprocessing/preprocessing_layer.py

```diff
@@ -0,0 +1,52 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_nlp.backend import config
+from keras_nlp.backend import keras
+from keras_nlp.backend import ops
+from keras_nlp.utils.tensor_utils import tensor_to_list
+
+
+class PreprocessingLayer(keras.layers.Layer):
+    """Preprocessing layer base class."""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._convert_input_args = False
+        self._allow_non_tensor_positional_args = True
+        self.built = True
+
+    def __call__(self, *args, **kwargs):
+        # Always place on CPU for preprocessing, to avoid expensive back and
+        # forth copies to GPU before the trainable model.
+        with tf.device("cpu"):
+            outputs = super().__call__(*args, **kwargs)
+
+            # Jax and Torch lack string tensors and a ragged type. We convert
+            # these outputs to simple python types (lists, strings) if
+            # 1) we are not running in the tensorflow backend, and
+            # 2) we are not running inside a tf.function (e.g.`tf.data`).
+            def convert_for_backend(x):
```

Comment:
Move this function out of the call definition (maybe `tensor_utils` is a good location)